### PR TITLE
[ADD] sale_order_line_sequence v10

### PR DESCRIPTION
**Pull request by serpentcs-dev1**
Sale Order Line with Sequence
========================

Displays the sequence of Sale order line and helps to maintain the order.

### CODE REVIEW

**Comment by serpentcs-dev1** [comment]
@aheficent Could you please review?

**Review by AaronHForgeFlow — COMMENTED** [review]
It works well, the creation and the modification of the SO and also when duplicating. Only one think. For previous SO the sequence is 10 for all the lines (See picture). I'm thinking about a pre-init hook to force update old sale orders in the database. Do you agree with me?
image

File: sale_order_line_sequence/model/sale_order.py
Comment (LoisRForgeFlow) [inline sale_order_line_sequence/model/sale_order.py]: Why not directly `return super(SaleOrder, self).with_context(keep_line_sequence=True).copy(default)`?

File: sale_order_line_sequence/model/sale_order.py
Comment (LoisRForgeFlow) [inline sale_order_line_sequence/model/sale_order.py]: is there any case in which this 9999 default would apply?? I cannot figure it out.

File: sale_order_line_sequence/model/sale_order.py
Comment (serpentcs-dev1) [reply sale_order_line_sequence/model/sale_order.py]: better, will change it!

File: sale_order_line_sequence/model/sale_order.py
Comment (serpentcs-dev1) [reply sale_order_line_sequence/model/sale_order.py]: If the whole Sale Order is copied the sequence would be kept as it is, but what would be the sequence if just the order line is duplicated? so its just sets the default sequence.

Comment (LoisRForgeFlow) [reply sale_order_line_sequence/model/sale_order.py]: But there is no way to functionally copy just lines, right? Anyway, this does not hurt. Keep it like this!

File: sale_order_line_sequence/model/sale_order.py
Comment (LoisRForgeFlow) [reply sale_order_line_sequence/model/sale_order.py]: To get rid of travis problem you can put it like this:


        return super(SaleOrder, self).with_context(
            keep_line_sequence=True).copy(default)

Comment (serpentcs-dev1) [reply sale_order_line_sequence/model/sale_order.py]: Yeah I did it like this, it works! Thanks!

        return super(SaleOrder,
                     self.with_context(keep_line_sequence=True)).copy(default)

**Review by LoisRForgeFlow — APPROVED** [review]
:+1:

File: sale_order_line_sequence/model/sale_order.py
Comment (serpentcs-dev1) [reply sale_order_line_sequence/model/sale_order.py]: Hmm.. I think you are right in a way, there is no need for the copy method.

When the line is copied it will trigger the create method which already resets and keeps the sequence.

I think we should remove it, what do you think?

Comment (LoisRForgeFlow) [reply sale_order_line_sequence/model/sale_order.py]: Yes, exactly. You can remove it since there are no test to break :wink:

**Review by AaronHForgeFlow — APPROVED** [review]
:+1: Thank you @darshan-serpent for attending all the suggestions. Can you please squash the commits before this is merged?

**Comment by AaronHForgeFlow** [comment]
It seems to me that the report is failing on runbot because the proper version of wkhtmltopdf is not installed :disappointed: 
image

**Comment by serpentcs-dev1** [comment]
@aheficent Yes encountered the same thing it is asking for upgrade, what is the solution to this as this could block the test for the report.

**Comment by AaronHForgeFlow** [comment]
Not sure if this can be configured at repo level. @moylop260 can you shed some light here?

**Comment by moylop260** [comment]
You can add a custom wkhtmltopdf version using the environment variable: `WKHTMLTOPDF_VERSION` from [.travis.yml](https://github.com/OCA/sale-workflow/blob/10.0/.travis.yml)

More info about: [travis_install_nightly](https://github.com/OCA/maintainer-quality-tools/blob/f54d6e33c5c48e6bd8051efe018e88ff3db669f8/travis/travis_install_nightly#L85)

**Comment by AaronHForgeFlow** [comment]
I think correct version is 0.12.2.1, isn't it?

**Comment by serpentcs-dev1** [comment]
Is it? I checked other PR but I cannot find any reference where there is WKHTMLTOPDF mentioned in the environment?

Okay I'll try applying 0.12.2.1

**Comment by serpentcs-dev1** [comment]
I don't think this is a proper way!!

I checked other PR, there the other reports are working fine, even though nothing is mentioned in their travis.yml file.

edit : changing to the way it was earlier.

**Comment by AaronHForgeFlow** [comment]
Not sure really how to fix this

**Comment by moylop260** [comment]
@aheficent I just apply the following patch in your branch:

```diff
diff --git a/.travis.yml b/.travis.yml
index 8cdb154aa..a22800348 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -13,11 +13,10 @@ addons:
       - python-simplejson
       - python-serial
       - python-yaml
-      - wkhtmltopdf
 
 env:
   global:
-  - VERSION="10.0" TESTS="0" LINT_CHECK="0" TRANSIFEX="0"
+  - VERSION="10.0" TESTS="0" LINT_CHECK="0" TRANSIFEX="0" WKHTMLTOPDF_VERSION="0.12.4"
   - TRANSIFEX_USER='transbot@odoo-community.org'
   - secure: mTXYvzW7/+xP465dNW0p9H+Q8jVgQXmqm8zo+ntZC6ESkTudj6xUkrjJcfzp9LGomxoG6N+6TezjFuhfGKCAc4S2manxj9YQoHvZMHJvt73VsMZU2QLfE+IMAdFS/AeBpYLNE8VvdKIGhiFwvHyRqaF0vAGwmcsOdS2NQb/hoiY=
 
```

If we add `- wkhtmltopdf` in `-addons -packages -apt` from `.travis.yml` file so a `apt-get install wkhtmltopdf` will be executed for ubuntu but default OS ubuntu version is trusty 14.04.
  - And use wkhtmltopdf 0.9.9: https://packages.ubuntu.com/trusty/wkhtmltopdf

If we remove that section of the .travis.yml runbot will work because our custom image have pre-installed a patched version.
But I added a explicit version in order to avoid a difference between travis and runbot for this package.

File: sale_order_line_sequence/model/sale_order.py
Comment (moylop260) [inline sale_order_line_sequence/model/sale_order.py]: What about use `self.env.context.get('keep_line_sequence')`? (if I use a `False` for this parameter)

File: sale_order_line_sequence/model/sale_order.py
Comment (moylop260) [inline sale_order_line_sequence/model/sale_order.py]: You can delete this section since that you are not changing the `default` variable.

**Review by moylop260 — APPROVED** [review]
LGTM
(Just small changes)

File: sale_order_line_sequence/model/sale_order.py
Comment (serpentcs-dev1) [reply sale_order_line_sequence/model/sale_order.py]: Right, for further use better to fetch the dict key's value, thanks!

**Review by AaronHForgeFlow — CHANGES_REQUESTED** [review]
The report doesn't look correct. The descriptions are missing in the report but exit in the SO lines

image

**Comment by serpentcs-dev1** [comment]
@aheficent Thanks for the review, this is fixed now!!

screenshot from 2017-07-06 10-56-25

**Review by AaronHForgeFlow — APPROVED** [review]
:+1: Report fine now. Please squash the commits before this is merged.

**Comment by serpentcs-dev1** [comment]
@aheficent  Squashed!!

**Pull request merged**
